### ISSUE TITLE
frida-trace crashed process on Samsung S22

### ISSUE THREAD

**Issue by tommylabunto** [frida/frida#2389]
Frida version: 16.0.2
Phone: Samsung S22
Android version: 13

When the command `frida-trace -i "recv*" -i "read*" -U -F` is used, it just crashed the app. Frida showed `Process crashed: Illegal instruction` and the following output is printed:

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
Revision: '13'
ABI: 'arm64'
Processor: '4'
Process uptime: 285s
Cmdline: com.facebook.katana
pid: 9034, tid: 9034, name: facebook.katana  >>> com.facebook.katana <<<
uid: 10276
tagged_addr_ctrl: 0000000000000001 (PR_TAGGED_ADDR_ENABLE)
pac_enabled_keys: 000000000000000f (PR_PAC_APIAKEY, PR_PAC_APIBKEY, PR_PAC_APDAKEY, PR_PAC_APDBKEY)
signal 4 (SIGILL), code 1 (ILL_ILLOPC), fault addr 0x00000075f0540654
    x0  00000000000000d8  x1  0000007ff2dcc1d0  x2  0000000000005460  x3  0000000000000040
    x4  0000000000000000  x5  0000000000000000  x6  0000000000000000  x7  000000006f9c9410
    x8  00000000000000cf  x9  0000000000000000  x10 0000000000000004  x11 00000000000000c0
    x12 0000007ff2dd1970  x13 000000007fffffff  x14 0000000000074170  x15 0000000618e66cff
    x16 00000075f0540654  x17 00000075e7a0200c  x18 00000075fd5f6000  x19 00000000000000d8
    x20 0000000000000064  x21 0000007ff2dd16ac  x22 0000007ff2dcc1d0  x23 0000000000005460
    x24 b4000072efc24d10  x25 0000007ff2dd16b8  x26 00000075fc7f9000  x27 0000000000000001
    x28 0000007ff2dd1990  x29 0000007ff2dcc140
    lr  00000075e7a0200c  sp  0000007ff2dcc140  pc  00000075f0540654  pst 0000000000001400
backtrace:
      #00 pc 00000000000ac654  /system/lib64/libgui.so (android::gui::BitTube::recvObjects(android::gui::BitTube*, void*, unsigned long, unsigned long)+72) 
      #01 pc 00000000000cf8ac  /system/lib64/libgui.so (android::DisplayEventDispatcher::processPendingEvents(long*, android::PhysicalDisplayId*, unsigned int*, android::gui::VsyncEventData*)+112) 
      #02 pc 00000000000cfd2c  /system/lib64/libgui.so (android::DisplayEventDispatcher::handleEvent(int, int, void*)+180)
      #03 pc 0000000000018028  /system/lib64/libutils.so (android::Looper::pollInner(int)+1064)
      #04 pc 0000000000017b9c  /system/lib64/libutils.so (android::Looper::pollOnce(int, int*, int*, void**)+116)
      #05 pc 00000000001687b0  /system/lib64/libandroid_runtime.so (android::android_os_MessageQueue_nativePollOnce(_JNIEnv*, _jobject*, long, int)+48) 
      #06 pc 00000000001c3964  /system/framework/arm64/boot-framework.oat (art_jni_trampoline+116) 
      #07 pc 000000000058d0e0  /system/framework/arm64/boot-framework.oat (android.os.MessageQueue.next+288) 
      #08 pc 000000000058a3e8  /system/framework/arm64/boot-framework.oat (android.os.Looper.loopOnce+104) 
      #09 pc 000000000058a2e0  /system/framework/arm64/boot-framework.oat (android.os.Looper.loop+560) 
      #10 pc 00000000002ee384  /system/framework/arm64/boot-framework.oat (android.app.ActivityThread.main+1316) 
      #11 pc 0000000000434600  /apex/com.android.art/lib64/libart.so (art_quick_invoke_static_stub+576)
      #12 pc 0000000000466d34  /apex/com.android.art/lib64/libart.so (_jobject* art::InvokeMethod<(art::PointerSize)8>(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, _jobject*, _jobject*, unsigned long)+1960) 
      #13 pc 0000000000466564  /apex/com.android.art/lib64/libart.so (art::Method_invoke(_JNIEnv*, _jobject*, _jobject*, _jobjectArray*) (.__uniq.165753521025965369065708152063621506277)+48) 
      #14 pc 0000000000099148  /system/framework/arm64/boot.oat (art_jni_trampoline+120)
      #15 pc 00000000005044c0  /system/framework/arm64/boot-framework.oat (com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run+144)
      #16 pc 000000000050d7dc  /system/framework/arm64/boot-framework.oat (com.android.internal.os.ZygoteInit.main+3404) 
      #17 pc 0000000000434600  /apex/com.android.art/lib64/libart.so (art_quick_invoke_static_stub+576) 
      #18 pc 000000000057e378  /apex/com.android.art/lib64/libart.so (art::JValue art::InvokeWithVarArgs<_jmethodID*>(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, _jmethodID*, std::__va_list)+900)
      #19 pc 00000000005f1d7c  /apex/com.android.art/lib64/libart.so (art::JNI<true>::CallStaticVoidMethodV(_JNIEnv*, _jclass*, _jmethodID*, std::__va_list)+160)
      #20 pc 00000000000c1c04  /system/lib64/libandroid_runtime.so (_JNIEnv::CallStaticVoidMethod(_jclass*, _jmethodID*, ...)+124) 

If `-F` is not specified, it crashed `system_server` and rebooted the phone. Anyone has issues running frida on S22?

**Comment by curioustwo** [comment]
I believe this is due to hardware mitigations called PAC/BTI that was introduced in ARMv8.5. It requires an indirect  branch to land on the "PAC" or "BTI" instruction. Frida will override the entry point to functions when it is instrumenting the code and is currently not handling the fact that the first instruction in the function prolog needs to be BTI/PAC. 

See:
[https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/armv8-1-m-pointer-authentication-and-branch-target-identification-extension](url)

@oleavr: is adding support for BTI/PAC enabled binaries on the roadmap for Frida? These instructions can be added unconditionally since they are a no-op on older ARMv8 designs.

**Comment by oleavr** [comment]
> I believe this is due to hardware mitigations called PAC/BTI that was introduced in ARMv8.5. It requires an indirect  branch to land on the "PAC" or "BTI" instruction. Frida will override the entry point to functions when it is instrumenting the code and is currently not handling the fact that the first instruction in the function prolog needs to be BTI/PAC. 
> 
> 
> 
> See:
> 
> [https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/armv8-1-m-pointer-authentication-and-branch-target-identification-extension](url)
> 
> 
> 
> @oleavr: is adding support for BTI/PAC enabled binaries on the roadmap for Frida? These instructions can be added unconditionally since they are a no-op on older ARMv8 designs.

Definitely. Would you be interested in taking a stab at a PR addressing this? 😊

**Comment by 350030173** [comment]
i think we have a same question :
[](https://github.com/frida/frida/issues/2275 )

**Comment by 350030173** [comment]
> > I believe this is due to hardware mitigations called PAC/BTI that was introduced in ARMv8.5. It requires an indirect  branch to land on the "PAC" or "BTI" instruction. Frida will override the entry point to functions when it is instrumenting the code and is currently not handling the fact that the first instruction in the function prolog needs to be BTI/PAC.
> > See:
> > [https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/armv8-1-m-pointer-authentication-and-branch-target-identification-extension](url)
> > @oleavr: is adding support for BTI/PAC enabled binaries on the roadmap for Frida? These instructions can be added unconditionally since they are a no-op on older ARMv8 designs.
> 
> Definitely. Would you be interested in taking a stab at a PR addressing this? blush

frida-gadget seems to hook sigaction. It may be hook wrong. Anyway, after loading frida-gadget, sigaction will get an error

**Comment by zjw88282740** [comment]
I write a simple fix, it works on my mi12 pro
https://gist.github.com/zjw88282740/b24ae6d913f8b323987d26cd6b38a8e7

**Comment by NgoHuy** [comment]
does the fix work for this issue?

**Comment by Manouchehri** [comment]
I'm still having this problem on an S24. Did ARMv9.2-A change something?

**Comment by radubogdan2k** [comment]
Did anyone else test [the fix above](https://gist.github.com/zjw88282740/b24ae6d913f8b323987d26cd6b38a8e7) and can confirm it has an effect?
